picard/cluster.py

```diff
@@ -443,183 +415,35 @@ def lookup_metadata(self):
             cluster.lookup_metadata()
 
 
-class ClusterDict(object):
-
+class FileCluster:
     def __init__(self):
-        # word -> id index
-        self.words = defaultdict(lambda: (-1, 0))
-        # id -> word, token index
-        self.ids = defaultdict(lambda: (None, None))
-        # counter for new id generation
-        self.id = 0
-        self.regexp = re.compile(r'\W', re.UNICODE)
-        self.spaces = re.compile(r'\s', re.UNICODE)
-
-    def get_size(self):
-        return self.id
-
-    def tokenize(self, word):
-        word = word.lower()
-        token = self.regexp.sub('', word)
-        return token if token else self.spaces.sub('', word)
-
-    def add(self, word):
-        """
-        Add a new entry to the cluster if it does not exist. If it
-        does exist, increment the count. Return the index of the word
-        in the dictionary or -1 is the word is empty.
-        """
-
-        if word == '':
-            return -1
-
-        index, count = self.words[word]
-        if index == -1:
-            token = self.tokenize(word)
-            if token == '':  # nosec
-                return -1
-            index = self.id
-            self.ids[index] = (word, token)
-            self.id = self.id + 1
-        self.words[word] = (index, count + 1)
-
-        return index
-
-    def get_word(self, index):
-        word, token = self.ids[index]
-        return word
+        self.files = []
+        self._artist_counts = defaultdict(lambda: 0)
+        self._artists = defaultdict(lambda: defaultdict(lambda: 0))
+        self._titles = defaultdict(lambda: 0)
 
-    def get_token(self, index):
-        word, token = self.ids[index]
-        return token
+    def add(self, album, artist, file):
+        self.files.append(file)
+        self._artist_counts[tokenize(artist)] += 1
+        self._artists[tokenize(artist)][artist] += 1
```

Comment:
I don't think we need `self._artist_counts` since we can use `len(self._artists[artist_token])` instead.

Reply:
It would actually be `sum(self._artists[artist_token])`. Not sure. We would first need to take the artist with highest sum, and then from that the individual highest "spelling variation". The separate artist_counts looks more efficient. But I'm not really happy about the implementation of the `artist` property. There must be a better way :D

Reply:
As far I understand this code, `self._artist_counts[tokenize(artist)] == len(self._artists[tokenize(artist)])` but that's not what you want it seems.
If you want artist full name having most occurrences in `self._artists[tokenize(artist)]`, that's `max(self._artists[tokenize(artist)])`.

Reply:
Still no :) It's a two step process of first choosing the most common artist, and then the most common spelling. We need an example. We have a cluster with different files and the following artist entries:

- artist 1
- ARTIST 1
- ARTIST 1
- artist1
- artist 2
- artist 2
- artist 2

Because of `tokenize` we have two entries in `_artists`: `artist1`, `artist2`.

`artist1` has 4 occurrences, `artist2` has 3. So we want to use `artist1`. That`s currently memorized in `_artist_count`.

`artist1` has 3 different spellings, "artist 1", "ARTIST 1" and "artist1". "ARTIST 1" is the most common, with 2 occurrences. That's what we want to use. These counts of different spellings for `artist1` are currently memorized in `self._artists["artist1"]`.

So to clarify: In the example `self._artist_counts[tokenize("artist1")] == 4`, but `len(self._artists[tokenize("artist1")]) == 3`. 

> If you want artist full name having most occurrences in self._artists[tokenize(artist)], that's max(self._artists[tokenize(artist)], key=self._artists[tokenize(artist)].get).

That's actually nicer then the itemgetter approach I used, so that means it could be:

```
@property
def artist(self):
    tokenized_artist = max(self._artist_counts, key=self._artist_counts.get)
    return max(self._artists[tokenized_artist], key=self._artists[tokenized_artist].get)

@property
def title(self):
    # Find the most common title
    return max(self._titles, key=self._titles.get)
```

Reply:
Yes, this is what I figured, see irc PM